Fix intersect1d crash with empty arrays

Harishjitu opened the pull request:
Fix illegal memory access when either input array is empty. The CUDA kernel was accessing non-existent memory, corrupting the CUDA context for subsequent operations.

Add early return with proper dtype handling using result_type() to match NumPy behavior.

Fixes #8937

leofang reviewed (changes requested):
Hi @Harishjitu Thanks for your contribution! I have two comments:
1. Could you please add a test, similar to [NumPy's](https://github.com/numpy/numpy/blob/c89cf2cd367bfa97e5c96ce90d3ae4534892a0c3/numpy/lib/tests/test_arraysetops.py#L36) and what's added in #8997, so that we capture this edge case?
2. Could you please try to avoid special-casing at the Python-level, and fix the kernel instead? @zazabap gave a patch in the #8997 PR description. I was able to discuss this with Copilot, which gave a similar patch (I did not verify, so please kindly do so before checking in the changes 🙂).

<details>
<summary>CLICK ME for the chat record</summary>

Here's the diff to fix the kernels in `cupy/_sorting/search.py`:

```diff
diff --git a/cupy/_sorting/search.py b/cupy/_sorting/search.py
index 7f0731d.. XXXXXXX 100644
--- a/cupy/_sorting/search. py
+++ b/cupy/_sorting/search.py
@@ -257,6 +257,11 @@ _hip_preamble = r'''
 
 _searchsorted_code = '''
     #ifdef __HIP_DEVICE_COMPILE__
     bool is_done = false;
     #endif
+
+    // Handle empty bins array to avoid illegal memory access
+    if (n_bins == 0) {
+        no_thread_divergence( y = 0 , true )
+    }
 
     // Array is assumed to be monotonically
@@ -364,10 +369,15 @@ _hip_preamble = r'''
 _exists_kernel = _core. ElementwiseKernel(
     'S x, raw T bins, int64 n_bins, bool invert',
     'bool out',
     '''
+    // Handle empty bins array to avoid illegal memory access
+    if (n_bins == 0) {
+        out = (invert ?  true : false);
+        return;
+    }
+
     const bool assume_increasing = true;
     const bool side_is_right = false;
     long long y;
     '''
     + _searchsorted_code + '''
@@ -378,10 +388,15 @@ _exists_kernel = _core. ElementwiseKernel(
 _exists_and_searchsorted_kernel = _core. ElementwiseKernel(
     'S x, raw T bins, int64 n_bins, bool invert',
     'bool out, int64 y',
     '''
+    // Handle empty bins array to avoid illegal memory access
+    if (n_bins == 0) {
+        out = (invert ? true : false);
+        y = 0;
+        return;
+    }
+
     const bool assume_increasing = true;
     const bool side_is_right = false;
     '''
     + _searchsorted_code + '''
```

**Key changes:**

1. **In `_searchsorted_code`** (line 258+): Added early return when `n_bins == 0` to set `y = 0` safely
2. **In `_exists_kernel`** (line 367+): Added guard to return `false` (or `true` if inverted) when `n_bins == 0`
3. **In `_exists_and_searchsorted_kernel`** (line 381+): Added guard to return both `out` and `y` safely when `n_bins == 0`

This way, **no changes are needed to `intersect1d` in `truth. py`** - the kernels handle empty arrays gracefully, and the algorithm naturally produces empty results with the correct dtype through array indexing operations like `arr1[mask]`.

This approach truly mirrors NumPy's philosophy where the algorithm handles edge cases implicitly rather than through explicit size checks at the Python level. 

</details>

Harishjitu commented:
Hi @leofang!
Thanks for the feedback! I've updated the PR:

1. **Fixed the kernels** in `cupy/_sorting/search.py`:
   - Added empty array guard to `_searchsorted_code`
   - Added empty array guard to `_exists_kernel`
   - Added empty array guard to `_exists_and_searchsorted_kernel`

2. **Added tests**:
   - `test_intersect1d_both_empty` — similar to NumPy's `intersect1d([], [])`
   - `test_intersect1d_empty_array` — similar to #8997 (first array empty)
   - `test_intersect1d_second_empty_array` — original bug case (second array empty)

Tested locally and all cases pass.

leofang commented on cupy/_logic/truth.py:
The whole point of modifying the kernels is to avoid special-casing at the Python level. Please undo the change in this file.

Harishjitu commented on cupy/_sorting/search.py:
```diff
@@ -260,6 +260,11 @@ def argwhere(a):
+    if (n_bins == 0) {
+        no_thread_divergence( y = 0 , true )
+    }
+
```
You are right!
_exists_kernel and _exists_and_searchsorted_kernel already return early before _searchsorted_code runs, so that if check was redundant

I removed that check

leofang reviewed (dismissed):
Thanks, @Harishjitu! I can confirm locally that the illegal instruction is fixed. However, there is one edge case where NumPy and CuPy (with this PR) disagrees in the returned array's dtype:
```python
>>> import cupy as cp
>>> import numpy as np
>>> cp.intersect1d(cp.array([0]), cp.array([]))
array([], dtype=int64)
>>> np.intersect1d(np.array([0]), np.array([]))
array([], dtype=float64)
```
I think it is likely due to the two inputs' dtypes differ (int64 vs float64), but I did not dig deep. If you have any thoughts, feel free to fix. Otherwise, I think this PR does fix the issue and I'd like to get it merged once we sort out the CI issues (discussing offline with the team, cc @kmaehashi for vis).

leofang commented:
/test mini
